// json-joy/esm/json-crdt/model/api/ModelApi.js
import { ArrayApi, BinaryApi, ConstApi, NodeApi, ObjectApi, StringApi, TupleApi, ValueApi } from './nodes';
import { PatchBuilder } from '../../../json-crdt-patch/PatchBuilder';
import { ArrayLww } from '../../types/lww-array/ArrayLww';
import { Const } from '../../types/const/Const';
import { ObjectLww } from '../../types/lww-object/ObjectLww';
import { ArrayRga } from '../../types/rga-array/ArrayRga';
import { BinaryRga } from '../../types/rga-binary/BinaryRga';
import { StringRga } from '../../types/rga-string/StringRga';
import { ValueLww } from '../../types/lww-value/ValueLww';
export class ModelApi {
    model;
    builder;
    next = 0;
    constructor(model) {
        this.model = model;
        this.builder = new PatchBuilder(this.model.clock);
    }
    changeQueued = false;
    queueChange = () => {
        if (this.changeQueued)
            return;
        this.changeQueued = true;
        queueMicrotask(() => {
            this.changeQueued = false;
            const et = this.et;
            if (et)
                et.dispatchEvent(new CustomEvent('change'));
        });
    };
    et = undefined;
    get events() {
        let et = this.et;
        if (!et) {
            this.et = et = new EventTarget();
            this.model.onchange = this.queueChange;
        }
        return et;
    }
    wrap(node) {
        if (node instanceof ValueLww)
            return node.api || (node.api = new ValueApi(node, this));
        else if (node instanceof StringRga)
            return node.api || (node.api = new StringApi(node, this));
        else if (node instanceof BinaryRga)
            return node.api || (node.api = new BinaryApi(node, this));
        else if (node instanceof ArrayRga)
            return node.api || (node.api = new ArrayApi(node, this));
        else if (node instanceof ObjectLww)
            return node.api || (node.api = new ObjectApi(node, this));
        else if (node instanceof Const)
            return node.api || (node.api = new ConstApi(node, this));
        else if (node instanceof ArrayLww)
            return node.api || (node.api = new TupleApi(node, this));
        else
            throw new Error('UNKNOWN_NODE');
    }
    get node() {
        return new NodeApi(this.model.root.node(), this);
    }
    get r() {
        return new NodeApi(this.model.root, this);
    }
    in(path) {
        return this.r.in(path);
    }
    find(path) {
        return this.node.find(path);
    }
    val(path) {
        return this.node.val(path);
    }
    tup(path) {
        return this.node.tup(path);
    }
    str(path) {
        return this.node.str(path);
    }
    bin(path) {
        return this.node.bin(path);
    }
    arr(path) {
        return this.node.arr(path);
    }
    obj(path) {
        return this.node.obj(path);
    }
    const(path) {
        return this.node.const(path);
    }
    root(json) {
        const builder = this.builder;
        builder.root(builder.json(json));
        this.apply();
        return this;
    }
    apply() {
        const ops = this.builder.patch.ops;
        const length = ops.length;
        const model = this.model;
        for (let i = this.next; i < length; i++)
            model.applyOperation(ops[i]);
        this.next = length;
        model.tick++;
        model.onchange?.();
    }
    advance() {
        this.next = this.builder.patch.ops.length;
        const model = this.model;
        model.tick++;
        model.onchange?.();
    }
    view() {
        return this.model.view();
    }
    flush() {
        const patch = this.builder.flush();
        this.next = 0;
        return patch;
    }
}
